pkg/pipelinerunmetrics/metrics.go

```diff
@@ -0,0 +1,151 @@
+package pipelinerunmetrics
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/jonboulle/clockwork"
+
+	pipelinev1beta1 "github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+	"github.com/tektoncd/results/pkg/apis/config"
+	"go.opencensus.io/stats"
+	"go.opencensus.io/stats/view"
+	"go.opencensus.io/tag"
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+	"knative.dev/pkg/apis"
+	"knative.dev/pkg/metrics"
+)
+
+var (
+	prDeleteCount        = stats.Int64("results_pipelinerun_delete_count", "total number of deleted pipelineruns", stats.UnitDimensionless)
+	prDeleteCountView    *view.View
+	prDeleteDuration     = stats.Float64("results_pipelinerun_delete_duration_seconds", "the pipelinerun deletion time in seconds", stats.UnitSeconds)
+	prDeleteDurationView *view.View
+	pipelineTag          = tag.MustNewKey("pipeline")
+	namespaceTag         = tag.MustNewKey("namespace")
+	statusTag            = tag.MustNewKey("status")
+)
+
+// Recorder is used to actually record PipelineRun metrics
+type Recorder struct {
+	clock clockwork.Clock
+}
+
+// NewRecorder creates a new metrics recorder instance
+// to log the PipelineRun related metrics
+func NewRecorder() *Recorder {
+	return &Recorder{clock: clockwork.NewRealClock()}
+}
+
+func viewRegister(logger *zap.SugaredLogger, cfg *config.Metrics) error {
+	var tags []tag.Key
+	switch cfg.PipelinerunLevel {
+	case config.PipelinerunLevelAtPipeline:
+		tags = []tag.Key{pipelineTag}
+	case config.PipelinerunLevelAtNS:
+		tags = []tag.Key{}
+	default:
+		return errors.New("invalid config for PipelinerunLevel: " + cfg.PipelinerunLevel)
+	}
+	prDeleteCountView = &view.View{
+		Description: prDeleteCount.Description(),
+		TagKeys:     []tag.Key{statusTag, namespaceTag},
+		Measure:     prDeleteCount,
+		Aggregation: view.Count(),
+	}
+
+	var distribution *view.Aggregation
+	switch cfg.DurationPipelinerunType {
+	case config.DurationPipelinerunTypeLastValue:
+		distribution = view.LastValue()
+	case config.DurationPipelinerunTypeHistogram:
+		distribution = view.Distribution(10, 30, 60, 300, 900, 1800, 3600, 5400, 10800, 21600, 43200, 86400)
```

Comment:
Not sure, do we need a 5s bucket? I am just trying to think if it will be useful.

Reply:
Not sure either. 
In normal cases users should see almost all delete duration in 10s/30s bucket. It would be unusual if it takes too long to delete the "finished" resource. 
So I think the bucket  distribution is not so important, people just need the bucket of small number and the bucket of the too-big number. 
We could modify the distribution if there are more use cases.